File: pxr/usd/ar/testenv/testArURIResolver.py
#!/pxrpythonsubst
#
# Copyright 2020 Pixar
#
# Licensed under the terms set forth in the LICENSE.txt file available at
# https://openusd.org/license.
#
import os
import unittest

# Force the use of ArDefaultResolver as the primary resolver for
# this test.
os.environ["PXR_AR_DISABLE_PLUGIN_RESOLVER"] = "1"

from pxr import Plug, Ar, Tf

class TestArURIResolver(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # Register test resolver plugins
        # Test plugins are installed relative to this script
        testRoot = os.path.join(
            os.path.dirname(os.path.abspath(__file__)), 'ArPlugins')

        pr = Plug.Registry()

        testURIResolverPath = os.path.join(
            testRoot, 'lib/TestArURIResolver*/Resources/')
        pr.RegisterPlugins(testURIResolverPath)

        testPackageResolverPath = os.path.join(
            testRoot, 'lib/TestArPackageResolver*/Resources/')
        pr.RegisterPlugins(testPackageResolverPath)

    def test_Setup(self):
        # Verify that our test plugin was registered properly.
        pr = Plug.Registry()
        self.assertTrue(pr.GetPluginWithName('TestArURIResolver'))
        self.assertTrue(Tf.Type.FindByName('_TestURIResolver'))
        self.assertTrue(Tf.Type.FindByName('_TestOtherURIResolver'))

        self.assertTrue(pr.GetPluginWithName('TestArPackageResolver'))
        self.assertTrue(Tf.Type.FindByName('_TestPackageResolver'))

    def test_Resolve(self):
        resolver = Ar.GetResolver()

        # The test URI resolver handles asset paths of the form "test:..."
        # and simply returns the path unchanged. We can use this to
        # verify that our test URI resolver is getting invoked.

        # These calls to Resolve should hit the default resolver and not
        # the URI resolver, and since these files don't exist we expect
        # Resolve would return ""
        self.assertEqual(resolver.Resolve("doesnotexist"), "")
        self.assertEqual(resolver.Resolve("doesnotexist.package[foo.file]"), "")

        # These calls should hit the URI resolver, which should return the
        # given paths unchanged.
        self.assertEqual(resolver.Resolve("test://foo"), 
                         "test://foo")
        self.assertEqual(resolver.Resolve("test://foo.package[bar.file]"), 
                         "test://foo.package[bar.file]")

        self.assertEqual(resolver.Resolve("test-other://foo"),
                         "test-other://foo")
        self.assertEqual(
            resolver.Resolve("test-other://foo.package[bar.file]"),
            "test-other://foo.package[bar.file]")

        # These calls should hit the URI resolver since schemes are
        # case-insensitive.
        self.assertEqual(resolver.Resolve("TEST://foo"),
                         "TEST://foo")
        self.assertEqual(resolver.Resolve("TEST://foo.package[bar.file]"),
                         "TEST://foo.package[bar.file]")

        self.assertEqual(resolver.Resolve("TEST-OTHER://foo"),
                         "TEST-OTHER://foo")
        self.assertEqual(
            resolver.Resolve("TEST-OTHER://foo.package[bar.file]"),
            "TEST-OTHER://foo.package[bar.file]")

    def test_InvalidScheme(self):
        resolver = Ar.GetResolver()
        invalid_underbar_path = "test_other:/abc.xyz"
        invalid_utf8_path = "test-π-utf8:/abc.xyz"
        invalid_numeric_prefix_path = "113-test:/abc.xyz"
        invalid_colon_path = "other:test:/abc.xyz"
        self.assertFalse(resolver.Resolve(invalid_underbar_path))
        self.assertFalse(resolver.Resolve(invalid_utf8_path))
        self.assertFalse(resolver.Resolve(invalid_numeric_prefix_path))
        self.assertFalse(resolver.Resolve(invalid_colon_path))

    def testGetRegisteredURISchemes(self):
        "Tests that all URI schemes for discovered plugins are returned"

        # Note: these are lifted from valid entries in the 
        # TestArURIResolver_plugInfo.json. In other environments there may be 
        # additional URI resolvers registered.
        expectedUriSchemes = ['test', 'test-other']
        actualUriSchemes = Ar.GetRegisteredURISchemes()

        for expectedUriScheme in expectedUriSchemes:
            self.assertTrue(expectedUriScheme in actualUriSchemes)

    def test_ResolveForNewAsset(self):
        resolver = Ar.GetResolver()

        # The test URI resolver handles asset paths of the form "test:..."
        # and simply returns the path unchanged. We can use this to
        # verify that our test URI resolver is getting invoked.

        # These calls to ResolveForNewAsset should hit the default resolver and
        # not the URI resolver and return some non-empty path. If this did
        # hit the URI resolver it would trip a TF_AXIOM.
        self.assertNotEqual(resolver.ResolveForNewAsset("doesnotexist"), "")
        self.assertNotEqual(
            resolver.ResolveForNewAsset("doesnotexist.package[foo.file]"), "")

        # These calls should hit the URI resolver, which should return the
        # given paths unchanged.
        self.assertEqual(
            resolver.ResolveForNewAsset("test://foo"), "test://foo")
        self.assertEqual(
            resolver.ResolveForNewAsset("test://foo.package[bar.file]"), 
            "test://foo.package[bar.file]")

        # These calls should hit the URI resolver since schemes are
        # case-insensitive.
        self.assertEqual(
            resolver.ResolveForNewAsset("TEST://foo"), "TEST://foo")
        self.assertEqual(
            resolver.ResolveForNewAsset("TEST://foo.package[bar.file]"), 
            "TEST://foo.package[bar.file]")

    def test_CreateContextFromString(self):
        resolver = Ar.GetResolver()

        # Exercise the CreateContextFromString(s) API in Python.
        # Since the _TestURIResolverContext object isn't wrapped to
        # Python these tests are a bit limited, but more extensive
        # tests are in the C++ version of this test.

        # CreateContextFromString with an empty URI scheme should
        # be equivalent to CreateContextFromString with no URI scheme.
        searchPaths = os.pathsep.join(["/a", "/b"])
        self.assertEqual(
            resolver.CreateContextFromString(searchPaths),
            resolver.CreateContextFromString("", searchPaths))

        # CreateContextFromStrings should be equivalent to creating an
        # Ar.ResolverContext combining the results of CreateContextFromString.
        self.assertEqual(
            Ar.ResolverContext(
                (resolver.CreateContextFromString("test", "context str"),
                 resolver.CreateContextFromString("", searchPaths))),
            resolver.CreateContextFromStrings([
                ("", searchPaths),
                ("test", "context str")
            ]))

if __name__ == '__main__':
    unittest.main()

File: pxr/usd/ar/testenv/testArURIResolver.cpp
//
// Copyright 2020 Pixar
//
// Licensed under the terms set forth in the LICENSE.txt file available at
// https://openusd.org/license.
//
#include "pxr/pxr.h"

#include "TestArURIResolver_plugin.h"

#include "pxr/usd/ar/defaultResolverContext.h"
#include "pxr/usd/ar/defineResolverContext.h"
#include "pxr/usd/ar/resolver.h"
#include "pxr/usd/ar/resolverContext.h"
#include "pxr/usd/ar/resolverContextBinder.h"

#include "pxr/base/arch/systemInfo.h"
#include "pxr/base/plug/plugin.h"
#include "pxr/base/plug/registry.h"
#include "pxr/base/tf/diagnosticMgr.h"
#include "pxr/base/tf/pathUtils.h"
#include "pxr/base/tf/setenv.h"
#include "pxr/base/tf/status.h"
#include "pxr/base/tf/stringUtils.h"
#include "pxr/base/tf/token.h"

PXR_NAMESPACE_USING_DIRECTIVE;

static void
SetupPlugins()
{
    // Set the preferred resolver to ArDefaultResolver before
    // running any test cases.
    ArSetPreferredResolver("ArDefaultResolver");

    // Register TestArURIResolver plugin. We assume the build system will
    // install it to the ArPlugins subdirectory in the same location as
    // this test.
    const std::string uriResolverPluginPath =
        TfStringCatPaths(
            TfGetPathName(ArchGetExecutablePath()),
            "ArPlugins/lib/TestArURIResolver*/Resources/") + "/";

    PlugPluginPtrVector plugins =
        PlugRegistry::GetInstance().RegisterPlugins(uriResolverPluginPath);
    
    TF_AXIOM(plugins.size() == 1);
    TF_AXIOM(plugins[0]->GetName() == "TestArURIResolver");
}

static void
TestResolveWithContext()
{
    ArResolver& resolver = ArGetResolver();

    // Verify that the context object is getting bound in the _TestURIResolver.
    // The test resolver simply appends the string in the context object
    // to the end of the given path when resolving.
    ArResolverContext ctx(_TestURIResolverContext("context"));
    ArResolverContextBinder binder(ctx);
    TF_AXIOM(resolver.Resolve("test://foo") == "test://foo?context");

    // Verify that binding another context overrides the previously-bound
    // context until the new binding is dropped.
    {
        ArResolverContext ctx2(_TestURIResolverContext("context2"));
        ArResolverContextBinder binder2(ctx2);
        TF_AXIOM(resolver.Resolve("test://foo") == "test://foo?context2");
    }
    TF_AXIOM(resolver.Resolve("test://foo") == "test://foo?context");

    // Verify that binding an unrelated context blocks the previously-bound
    // context.
    {
        ArResolverContext ctx3(ArDefaultResolverContext{});
        ArResolverContextBinder binder3(ctx3);
        TF_AXIOM(resolver.Resolve("test://foo") == "test://foo");
    }
    TF_AXIOM(resolver.Resolve("test://foo") == "test://foo?context");
}

static void
TestCreateContextFromString()
{
    ArResolver& resolver = ArGetResolver();

    const std::vector<std::string> searchPaths = { "/a", "/b" };
    const std::string searchPathStr = 
        TfStringJoin(searchPaths, ARCH_PATH_LIST_SEP);

    // CreateContextFromString with an empty URI scheme should be
    // equivalent to CreateContextFromString without a URI scheme.
    TF_AXIOM(
        resolver.CreateContextFromString("", searchPathStr) ==
        ArResolverContext(ArDefaultResolverContext(searchPaths)));

    TF_AXIOM(
        resolver.CreateContextFromString("", searchPathStr) ==
        resolver.CreateContextFromString(searchPathStr));

    // CreateContextFromString with a URI scheme that has no registered
    // resolver results in an empty ArResolverContext.
    TF_AXIOM(
        resolver.CreateContextFromString("bogus", "context string") ==
        ArResolverContext());
    
    // CreateContextFromString with a URI scheme with a registered resolver
    // results in whatever context is returned from that resolver.
    TF_AXIOM(
        resolver.CreateContextFromString("test", "context string") ==
        ArResolverContext(_TestURIResolverContext("context string")));

    // CreateContextFromStrings should return a single ArResolverContext
    // containing context objects based on the given URI schemes and
    // context strings.
    TF_AXIOM(
        resolver.CreateContextFromStrings({
            {"test", "context string"}
        }) ==
        ArResolverContext(
            _TestURIResolverContext("context string")
        ));

    TF_AXIOM(
        resolver.CreateContextFromStrings({
            {"", TfStringJoin(searchPaths, ARCH_PATH_LIST_SEP)},
            {"test", "context string"},
            {"bogus", "context string" }
        }) ==
        ArResolverContext(
            ArDefaultResolverContext(searchPaths),
            _TestURIResolverContext("context string")));
}

static void
TestCreateDefaultContext()
{
    ArResolver& resolver = ArGetResolver();

    // CreateDefaultContext should return an ArResolverContext containing
    // the union of the default contexts returned by all registered resolvers.
    // ArDefaultResolver returns an empty ArResolverContext object as its
    // default so we can't test for that, but TestArURIResolver returns a
    // _TestURIResolverContext which we can check for here.
    const ArResolverContext defaultContext = resolver.CreateDefaultContext();

    const _TestURIResolverContext* uriCtx =
        defaultContext.Get<_TestURIResolverContext>();
    TF_AXIOM(uriCtx);
    TF_AXIOM(uriCtx->data == "CreateDefaultContext");
}

static void
TestCreateDefaultContextForAsset()
{
    auto runTest = [](const std::string& assetPath) {
        ArResolver& resolver = ArGetResolver();

        // CreateDefaultContextForAsset should return an ArResolverContext
        // containing the union of the default contexts returned by all
        // registered resolvers.
        const ArResolverContext defaultContext = 
            resolver.CreateDefaultContextForAsset(assetPath);

        // ArDefaultResolver returns an ArDefaultResolverContext whose search
        // path is set to the directory of the asset.
        {
            const ArDefaultResolverContext* defaultCtx =
                defaultContext.Get<ArDefaultResolverContext>();
            TF_AXIOM(defaultCtx);

            const ArDefaultResolverContext expectedCtx(
                std::vector<std::string>{ 
                    TfGetPathName(TfAbsPath(assetPath)) });
            TF_AXIOM(*defaultCtx == expectedCtx);
        }

        // TestArURIResolver returns a _TestURIResolverContext whose data field
        // is set to the absolute path of the asset.
        {
            const _TestURIResolverContext* uriCtx =
                defaultContext.Get<_TestURIResolverContext>();
            TF_AXIOM(uriCtx);
        
            const _TestURIResolverContext expectedCtx(
                TfAbsPath("test/test.file"));
            TF_AXIOM(*uriCtx == expectedCtx);
        }
    };

    runTest("test/test.file");

    // For a package-relative path, CreateDefaultContextForAsset should only
    // consider the outer-most package path.
    runTest("test/test.file[in_package]");
}

int main(int argc, char** argv)
{
    SetupPlugins();

    printf("TestResolveWithContext ...\n");
    TestResolveWithContext();

    printf("TestCreateContextFromString ...\n");
    TestCreateContextFromString();

    printf("TestCreateDefaultContext ...\n");
    TestCreateDefaultContext();

    printf("TestCreateDefaultContextForAsset ...\n");
    TestCreateDefaultContextForAsset();

    printf("Test PASSED\n");
    return 0;
}
